### PR TITLE
CLI: explicit integration with cgo

### PR DESCRIPTION
**Pull request by andrewrk**
This is just a proof-of-concept. Maybe it could be nice for Go
developers, but also maybe we don't put additional complexity into the
zig codebase when CLI flags work just fine.

See the discussion in #7342

### CODE REVIEW

File: src/target.zig [hunk 1]
```diff
@@ -343,3 +343,79 @@ pub fn is_libcpp_lib_name(target: std.Target, name: []const u8) bool {
...
+    } else if (std.mem.eql(u8, go_os, "windows")) {
+        os_tag = .windows;
+        abi = .gnu; // so we can provide libc
+    } else {
+        return error.UnrecognizedOperatingSystem;
+    }
+
+    var cpu_arch: ?std.Target.Cpu.Arch = undefined;
+    if (std.mem.eql(u8, go_arch, "386")) {
+        cpu_arch = null;
```
Comment (Vexu) [inline src/target.zig:387]:
```suggestion
    if (std.mem.eql(u8, go_arch, "native")) {
        cpu_arch = null;
```

**Comment by lu4p** [comment]
I think this would be great, to onboard members of the go community to the zig project.

File: src/target.zig [hunk 2]
```diff
@@ -343,3 +343,79 @@ pub fn is_libcpp_lib_name(target: std.Target, name: []const u8) bool {
...
+        os_tag = .dragonfly;
+    } else if (std.mem.eql(u8, go_os, "freebsd")) {
+        os_tag = .freebsd;
+    } else if (std.mem.eql(u8, go_os, "illumos")) {
+        return error.UnsupportedOperatingSystem;
+    } else if (std.mem.eql(u8, go_os, "js")) {
+        return error.UnsupportedOperatingSystem;
+    } else if (std.mem.eql(u8, go_os, "linux")) {
+        os_tag = .linux;
+        abi = .musl; // go wants static linking on linux
```
Comment (lu4p) [inline src/target.zig:369]: I think this is an ok default behaviour, but this should be configurable via a flag or env var.

**Comment by LemonBoy** [comment]
I don't think that writing a shell script to perform the argument translation is _that_ hard.
On the other hand maintaining this set of extra switches and making sure they don't regress is much harder.

File: src/main.zig [hunk 3]
```diff
@@ -154,6 +149,51 @@ pub fn mainArgs(gpa: *Allocator, arena: *Allocator, args: []const []const u8) !v
         }
     }
 
+    const is_cgo = std.os.getenvZ("ZIG_CGO") != null;
+    if (is_cgo) cgo: {
+        // Workaround for https://github.com/golang/go/issues/43078, here we fix the order of
+        // command line arguments. Dear Go developers please let the Zig project know when
+        // we can remove this!
```
Comment (komuw) [inline src/main.zig:156]: https://go-review.googlesource.com/c/go/+/276412/ which is under review may fix this.

**Comment by jonjohnsonjr** [comment]
> I don't think that writing a shell script to perform the argument translation is that hard.
> On the other hand maintaining this set of extra switches and making sure they don't regress is much harder.

I tend to agree, though I don't personally want to be responsible for maintaining this mapping (and it seems silly to have everyone individually reimplement it), but it's not clear where something like that _should_ live.

I'm not sure what the minimum bar is for creating a repository is under the ziglang org, but something like a github.com/ziglang/zig-cgo (or cgo-zig) that hosts the canonical translation layer would be a very nice thing to be able to reference. There seems to be precedent for this with docker-zig, zig.vim, vscode-zig, zig-mode, etc...

**Comment by andrewrk** [comment]
The more I think about it, the more I think there should not be a translation layer at all - typically when you are creating a cross build, you have in mind the target that you want to specify. You will then have the task of choosing a `-target` / `-mcpu` flags for Zig and GOOS/GOARCH env vars for go. It's possible that the target you want to give to Zig can't even be described by GOOS/GOARCH. For example OS version ranges, CPU features, dynamic linker path, C ABI, glibc version. In this case you *must* manually choose the Zig target configuration. So I think we may actually be doing a disservice by leading users to configure the much more detailed Zig target configuration via the coarse-grained go env vars.

Based on this, and @LemonBoy's point, I'm going to close this. I do, however, want to solve the problem of having zig choose a more effective local cache directory. (See https://github.com/golang/go/issues/43078#issuecomment-741198473 for more details on that.)

**Pull request closed without merging**